src/net_processing.cpp

```diff
@@ -1642,37 +1645,31 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
     const std::chrono::seconds mempool_req = pfrom->m_tx_relay != nullptr ? pfrom->m_tx_relay->m_last_mempool_req.load()
                                                                           : std::chrono::seconds::min();
 
-    {
-        LOCK(cs_main);
-
-        // Process as many TX items from the front of the getdata queue as
-        // possible, since they're common and it's efficient to batch process
-        // them.
-        while (it != pfrom->vRecvGetData.end() && (it->type == MSG_TX || it->type == MSG_WITNESS_TX)) {
-            if (interruptMsgProc)
-                return;
-            // The send buffer provides backpressure. If there's no space in
-            // the buffer, pause processing until the next call.
-            if (pfrom->fPauseSend)
-                break;
+    // Process as many TX items from the front of the getdata queue as
+    // possible, since they're common and it's efficient to batch process
+    // them.
+    while (it != pfrom->vRecvGetData.end() && (it->type == MSG_TX || it->type == MSG_WITNESS_TX)) {
```

Comment:
You are no longer holding cs_main. Why is it OK to access vRecvGetData here without a lock?

Reply:
`vRecvGetData` isn't guarded by cs_main. In fact, it's not guarded by anything 😱 

All access of `vRecvGetData` is in `ProcessMessages()`, which only ever runs single threaded in the Message Handler thread, so this is safe, but ideally it'd be guarded by some synchronization method in case that ever changes.

Reply:
Right. Previous to this commit, the `mempool`, `vRecvGetData`, and `m_tx_relay` were all accessed atomically under `cs_main`. This changes that and uses finer-grained locking. 

I think all the other accesses are safe but I wasn't certain about `vRecvGetData`.

Reply:
@narula Nice catch. Indeed, as @jnewbery says, it isn't formally protected by anything right now. That's fine because (a) it's only accessed from one thread (b) every access to it actually holds `cs_vRecv` (just checked).

Reply:
Could you point out where `cs_vRecv` is taken before `vRecvGetData` is accessed here? I only see it locked in two places in net.cpp, and I can't connect either of them to `ProcessMessages`. The non-stats one is in `ReceiveMsgBytes`.

Reply:
net locks `cs_vRecv` and holds it while calling ProcessMessages. I checked by adding the annotation and recompiling.

Reply:
ok! based on discussion in PR review club today I'll PR a change to guard it by `cs_vRecv`, and look at `orphan_work_set` too. To be clear: there's no data race now, but it's worth fixing up in case these fields are accessed from multiple threads in the future.